Add H csr vsip,hgeie,hideleg,hie,hip,hvip,vsie,vsscratch,hgeip

Shehrozkashif opened the pull request:
This pr includes  csr  vsip,hegeie,hideleg,hie,hip,hvip,vsie,vsscratch,hgeip

Shehrozkashif committed:
feat(csr): add H-mode CSRs vsip, hegeie, hideleg, hie, hip, hvip, vsie, vsscratch, hgeip

Shehrozkashif committed:
feat(csr): add H-mode CSRs vsip, hegeie, hideleg, hie, hip, hvip, vsie, vsscratch, hgeip

ThinkOpenly commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,58 @@
+    reset_value: 0
+    description: Virtual Supervisor Software Interrupt.
+
+  STIP:
```
`STI`

ThinkOpenly commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,58 @@
+    location: 5
+    type: RO
+    reset_value: 0
+    description: Standard S-level interrupt. Read-only zero.
```
Supervisor Timer Interrupt

ThinkOpenly commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,58 @@
+    reset_value: 0
+    description: Virtual Supervisor Timer Interrupt.
+
+  SEIP:
```
`SEI`

ThinkOpenly commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,58 @@
+    location: 9
+    type: RO
+    reset_value: 0
+    description: Standard S-level interrupt. Read-only zero.
```
"Supervisor External Interrupt"

ThinkOpenly commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,58 @@
+    description: Virtual Supervisor External Interrupt.
+
+  LCOFIP:
+    location: 13
```
`12`

ThinkOpenly replied:
I see you deleted this one, which was not even supposed to be here. OK

ThinkOpenly commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,58 @@
+    reset_value: 0
+    description: Standard S-level interrupt. Read-only zero.
+
+  VSTIP:
```
`VSTI`

ThinkOpenly commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,58 @@
+    reset_value: 0
+    description: Standard S-level interrupt. Read-only zero.
+
+  VSEIP:
```
`VSEI`

ThinkOpenly reviewed (changes requested):
I did not finish the review, but thought this feedback would be helpful for now.

Shehrozkashif commented on arch/csr/H/hgeie.yaml:
```diff
@@ -0,0 +1,42 @@
+name: hgeie
+long_name: Hypervisor Guest External Interrupt Enable Register
+description: |
+  The `hgeie` register is an HSXLEN-bit read/write register that contains enable bits
+  for the guest external interrupts at this hart. Each bit in `hgeie` corresponds to
+  a guest external interrupt and determines whether that interrupt is enabled.
```
Got it, I’ll fix it to be more like the spec.

ThinkOpenly replied:
Restore backticks around all the CSRs in all the description fields.

If you could make these LONG lines shorter (less than about 70-80 characters), that is preferred.

ThinkOpenly commented on arch/csr/H/vsie.yaml:
```diff
@@ -0,0 +1,58 @@
+  normally read or modify sie actually access vsie instead. However, interrupts directed to
+  HS-level continue to be indicated in the HS-level sip register, not in vsip, when V=1.
+
+  The standard portion (bits 15:0) of vsie is formatted as follows:
+  15 14 13 12 10 9 8 6 5 4 2 1 0
+  0 LCOFIE 0 SEIE 0 STIE 0 SSIE 0
+  2 1 3 1 3 1 3 1 1
+
```
This is obviously not formatted well. Delete it. The fields information below covers the same information.

Shehrozkashif commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,57 @@
+    location: 2
+    type: RW
+    reset_value: 0
+    description: Virtual Supervisor Software Interrupt.
```
description is required for CSR fields and cannot be replaced with long_name due to the schema structure. Using long_name results in the error: Property long_name is not allowed.

ThinkOpenly replied:
Not sure how "long_name" is not allowed, since `arch/csr/hstatus.yaml` uses it:
```
# yaml-language-server: $schema=../../schemas/csr_schema.json

$schema: "csr_schema.json#"
kind: csr
name: hstatus
long_name: Hypervisor Status
address: 0x600
priv_mode: S
length: SXLEN
description: |
  The hstatus register tracks and controls a VS-mode guest.

  Unlike fields in `sstatus`, which are all aliases of fields `mstatus`,
  bits in `hstatus` are independent bits and do not have aliases.
definedBy: H
fields:
  VSXL:
    long_name: VS-mode XLen
    location: 33-32
    base: 64
    description:
[...]
```

It is true that "description" is required, though. In the absence of a better/longer description, could you duplicate that text to both attributes?
```
    long_name: Virtual Supervisor Software Interrupt
    description: Virtual Supervisor Software Interrupt
```

dhower-qc replied:
long_name for CSR fields was (very belatedly) added in #598. It's now in main, so a rebase/merge should pick it up.

Shehrozkashif replied:
sure ill duplicate the text

Shehrozkashif replied:
I have done a rebase and now its working thanks !

Shehrozkashif commented on arch/csr/H/hie.yaml:
```diff
@@ -0,0 +1,53 @@
+      to be processed, based on the configuration in `hvip`.
+
+  reserved:
+    location: 13-15
```
well do I have to replicate separately the reserved bits like reserved_15-13 to 1-0  because in location field Iam unable to write "15-13,11,9-7,5-3,1-0" because of json error Incorrect type. Expected "number" ?

ThinkOpenly replied:
Ah. We apparently cannot represent CSR fields like we do opcode fields. CSR fields must be contiguous:
```
    "field_location": {
      "oneOf": [
        { "type": "number", "description": "Location of a single bit" },
        {
          "type": "string",
          "pattern": "^[0-9]+-[0-9]+$",
          "description": "Location range of a multi-bit field"
        }
      ],
```

Maybe we should do this differently: don't define the reserved fields at all, then enforce the "read-only zero" constraints in a "sw_write()" attribute. Does that make sense?

dhower-qc replied:
> don't define the reserved fields at all

That's the answer. Anything missing is treated as reserved-read-only-0.

Shehrozkashif commented on arch/csr/H/hip.yaml:
```diff
@@ -0,0 +1,46 @@
+length: SXLEN
+fields:
+  SGEIP:
+    location: 0
```
I think I was not on the right document before now I have got the right one apology for it !

dhower-qc commented on arch/csr/H/hvip.yaml:
```diff
@@ -0,0 +1,44 @@
+      Pending interrupt bit for VS-level software interrupts. This bit is writable and
+      is set to 1 to assert a VS-level software interrupt.
+
+  reserved_6_15:
```
We don't need to list reserved bits at all. If there is no field at a bit, it is assumed to be reserved

dhower-qc commented on arch/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,64 @@
+    type: RW
+    reset_value: 0
+    long_name: Supervisor Guest External Interrupt
+    description: Supervisor Guest External Interrupt
```
There is also a Local Counter Overflow Interrupt (LCOFI) at bit 13. It is optional whether or not the bit is writable. As such, we need configuration-dependent type and reset_value for the field, and a new parameter:

```yaml
LCOFI:
  location: 13
  definedBy: Sscofpmf
  long_name: Local Counter Overflow Interrupt
  description: Delegate Local Counter Overflow Interrupts to VS-mode.
  type(): return HIDELEG_LCOFI_MUTABLE ? CsrFieldType::RW : CsrFieldType::RO;
  reset_value(): return HIDELEG_LCOFI_MUTABLE ? UNDEFINED_LEGAL : 0;
```

in H.yaml:

```yaml
# ...
parameters:
  # ...
  HIDELEG_LCOFI_MUTABLE:
    schema:
      type: boolean
    description: When true, `hideleg`[13] is writable, and Local Counter Overflow Interrupts can be delegated to VS-mode.
```

dhower-qc commented on arch/csr/H/hie.yaml:
```diff
@@ -0,0 +1,46 @@
+definedBy: H
+length: SXLEN
+fields:
+  SGEIE:
```
Also needs LCOFI

dhower-qc commented on arch/csr/H/hie.yaml:
```diff
@@ -0,0 +1,46 @@
+# yaml-language-server: $schema=../../../schemas/csr_schema.json
```
all reset values are UNDEFINED_LEGAL

dhower-qc commented on arch/csr/H/hip.yaml:
```diff
@@ -0,0 +1,46 @@
+priv_mode: S
+definedBy: H
+length: SXLEN
+fields:
```
Also needs LCOFI

dhower-qc commented on arch/csr/H/hip.yaml:
```diff
@@ -0,0 +1,46 @@
+
+  VSTIP:
+    location: 6
+    type: RO
```
Should be RO-H

dhower-qc reviewed (changes requested):
This is a complicated set of CSRs. I've added a bunch of comments, but note that we're likely to need multiple review passes before it's ready.

Shehrozkashif commented on arch/csr/H/hip.yaml:
```diff
@@ -0,0 +1,46 @@
+      This bit is the logical OR of `vseip` from `hvip`, the interrupt from `hgeip` selected by `hstatus.VGEIN`,
+      and any other external interrupt signal directed to VS-level.
+
+  VSTIP:
+    location: 6
+    type: RO
+    reset_value: 0
+    description: |
+      Pending interrupt bit for VS-level timer interrupts (VSTI).
+      This bit is the logical OR of `vstip` from `hvip` and any other timer interrupt directed to VS-level.
```
sure ill add Sstc CSRs I believe these are they ?

16. "Sstc" Extension for Supervisor-mode Timer Interrupts, Version 1.0.0 123
16.1. Machine and Supervisor Level Additions 123
16.1.1. Supervisor Timer Register (stimecmp) 123
16.1.2. Machine Interrupt Registers (mip and mie) 124
16.1.3. Supervisor Interrupt Registers (sip and sie) 124
16.1.4. Machine Counter-Enable Register (mcounteren) 124
16.2. Hypervisor Extension Additions 124
16.2.1. Virtual Supervisor Timer Register (vstimecmp) 124
16.2.2. Hypervisor Interrupt Registers (hvip, hip, and hie) 125
16.2.3. Hypervisor Counter-Enable Register (hcounteren) 125
16.3. Environment Config (menvcfg/henvcfg) Support

Shehrozkashif replied:
Property sw_read() is not allowed.
getting schema error

AFOliveira replied:
@Shehrozkashif with just that info it's hard to give any meaningful input, can you share the sholw section where that happens?

ThinkOpenly replied:
`sw_read()` is a CSR property, not a CSR field property. Maybe that's why you're seeing a problem?

Shehrozkashif replied:
yes  paul is right! I can define this function  sw_write(csr_value) but while adding sw_read():(the function derek wants me to add ) is not in fields
sw_read() so I think I have to add it as a property not in fields ?

ThinkOpenly replied:
> have to add [`sw_read()`] as a property not in fields ?

Correct. It should be in the YAML at the "top level" as a peer with other attributes like "name", "description", and (not to be too confusing) "fields".

Shehrozkashif replied:
Got it, I’ll move sw_read() to the top level alongside name, description, and fields.

Shehrozkashif commented on arch/csr/H/hgeie.yaml:
```diff
@@ -0,0 +1,30 @@
+  GEI_ENABLE:
+    location_rv32: 31-1
+    location_rv64: 63-1
+    type: RW
```
can we write it in description ?

ThinkOpenly commented on arch/csr/H/hgeie.yaml:
```diff
@@ -0,0 +1,68 @@
+      return NUM_EXTERNAL_GUEST_INTERRUPTS > 0 ? CsrFieldType::RW : CsrFieldType::RO;
+    reset_value: UNDEFINED_LEGAL
+    sw_write(csr_value): |
+      return ((1 << NUM_EXTERNAL_GUEST_INTERRUPTS) - 1) & csr_value.GEI_ENABLE;
```
Derek suggested using ``MXLEN`1`` instead of `1` in the first instance.

ThinkOpenly commented on arch/csr/H/hgeip.yaml:
```diff
@@ -0,0 +1,59 @@
+    type: RO
+    reset_value: UNDEFINED_LEGAL
+    description: |
+      Enable bits; see description of hgeip
```
These are "Pending bits" (not "Enable bits").

ThinkOpenly commented on arch/csr/H/vsie.yaml:
```diff
@@ -0,0 +1,71 @@
+    reset_value: UNDEFINED_LEGAL
+    alias: hie.VSSIE[0]
+    description: |
+      SSIE. Read-only zero when hideleg[2] is 0. Else, alias of hie.VSSIE.
```
I imagine we need `sw_write` and `sw_read` to handle dependence on the value of `hideleg[2]`.

ThinkOpenly commented on arch/csr/H/vsip.yaml:
```diff
@@ -0,0 +1,73 @@
+      *Supervisor Software Interrupt Pending*
+
+      Indicates a pending software interrupt at the supervisor level.
+      Read-only zero when `hideleg[2] == 0`, else aliased to `hip.VSSIP[0]`.
```
Looks like we need `sw_read`/`sw_write` here, too.

ThinkOpenly commented on arch/csr/H/vsscratch.yaml:
```diff
@@ -0,0 +1,24 @@
+priv_mode: VS
+length: VSXLEN
+definedBy: H
+fields: {}
```
This should look like `sscratch.yaml` here, with a single SCRATCH field.

ThinkOpenly commented on arch/csr/H/vsie.yaml:
```diff
@@ -0,0 +1,80 @@
+    alias: hie.VSSIE[0]
+    description: |
+      SSIE. Read-only zero when hideleg[2] is 0. Else, alias of hie.VSSIE.
+      sw_read(): |
```
The `sw_read()` attribute is implemented as the top-level (peered with `definedBy`, `length`, `fields`, etc.), so should have no indentation.

ThinkOpenly commented on arch/csr/H/vsie.yaml:
```diff
@@ -0,0 +1,80 @@
+          } else {
+            return 0;
+          }
+      sw_write(csr_value): |
```
`sw_write` should be peered with the field attributes, like `location`, `type`, `reset_value`, etc. As it stands, it's getting included with the "description" data.

Shehrozkashif replied:
for sw_write Iam getting error Property sw_read() is not allowed.yaml-schema: CSR description(513)

ThinkOpenly replied:
You need the "function form" for these attribute names:
```
sw_read(): |
[...]
```
(Note the parentheses `()`.)

`sw_write` needs parentheses, too, and it should hierarchically under `fields/[field-name]`:
```
fields:
  SSIE:
    location: 1
[...]
    sw_write(): |
[...]
```

ThinkOpenly commented on arch/csr/H/vsie.yaml:
```diff
@@ -0,0 +1,104 @@
+    alias: hie.VSSIE[0]
+    description: |
+      SSIE. Read-only zero when hideleg[2] is 0. Else, alias of hie.VSSIE.
+    "sw_write(csr_value)": |
```
Remove the quotation marks around all of the `sw_write` to be consistent with the other CSR descriptions.

ThinkOpenly commented on arch/csr/H/hip.yaml:
```diff
@@ -0,0 +1,46 @@
+# yaml-language-server: $schema=../../../schemas/csr_schema.json
```
I presume this still needs to be added.

ThinkOpenly commented on arch/csr/H/vsscratch.yaml:
```diff
@@ -0,0 +1,30 @@
+  SCRATCH:
+    type: RW
+    reset_value: 0
+    location: 0
```
The SCRATCH field is the full register width, so you need to split it into two attributes, one for each supported length:
```suggestion
    location_rv32: 31-0
    location_rv64: 63-0
```

Shehrozkashif committed:
Update arch/csr/H/vsscratch.yaml

Co-authored-by: Paul Clarke <ThinkOpenly@gmail.com>
Signed-off-by: Shehroz Kashif <131602772+Shehrozkashif@users.noreply.github.com>

Shehrozkashif committed:
Update arch/csr/H/vsscratch.yaml

Co-authored-by: Paul Clarke <ThinkOpenly@gmail.com>
Signed-off-by: Shehroz Kashif <131602772+Shehrozkashif@users.noreply.github.com>

dhower-qc commented:
smoke fails the vsie sw_read function because you have Sail syntax in an IDL block:

```
sw_read(): |
  let value = 0;
```

should be

```
sw_read(): |
  Bits<64> value = 0;
```

Shehrozkashif committed:
feat(cpu): add Sstc CSRs

Shehrozkashif committed:
feat(cpu): add Sstc CSRs

dhower-qc commented on arch/csr/H/vsie.yaml:
```diff
@@ -0,0 +1,104 @@
+
+sw_read(): |
+  Bits<64> value = 0;
+  if (CSR[hideleg][2]) {
```
needs to convert to Bits:

```
  if ($bits(CSR[hideleg])[2]) {
```

multiple occurances of this.

ThinkOpenly commented:
Referencing #565. There appears to be a parallel effort to address this same issue.

Shehrozkashif committed:
Merge branch 'main' into H_CSRs

Shehrozkashif committed:
Merge branch 'main' into H_CSRs

ThinkOpenly commented:
How shall we proceed here? @Shehrozkashif, do you want to finish this? Do we need to break it up any? I've lost context on what else is needed here. The CI tests are certainly not happy.

Shehrozkashif committed:
Merge branch 'main' into H_CSRs

Shehrozkashif commented:
@ThinkOpenly I’d definitely be interested in seeing this resurrected. What do you think would be the best way to bring it back?

Shehrozkashif committed:
Merge branch 'riscv:main' into H_CSRs

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

Shehrozkashif committed:
fix(csr): address review comments for H-extension and Sstc CSRs

- Convert definedBy: H to new extension format across all new CSRs
- Add missing writable field to hgeie, hgeip, hideleg, hie, hip, hvip, vsscratch
- Replace schema-removed alias syntax with sw_read()/sw_write() IDL functions
- Fix vsie/vsip sw_read to correctly shift bits (hie bit 2->vsie bit 1, etc.)
- Fix vsie/vsip sw_write: use csr_value.FIELD instead of undefined value variable
- Fix hideleg bit indices in sw_write conditions to use delegation bits not field bits
- Add hip sw_read() computing SGEIP/VSEIP/VSTIP/VSSIP with Sstc support
- Convert vsscratch/stimecmp/vstimecmp tagged-text descriptions to plain text
- Add requires: Sstc and correct length to vstimecmp

Shehrozkashif committed:
Merge origin/H_CSRs: keep fixed CSRs, restore golden file

Shehrozkashif committed:
fix(csr): remove trailing blank line in vsscratch.yaml

Shehrozkashif committed:
fix(csr): rewrite hip sw_read() using valid IDL syntax

Replace C-style bool/unsigned int with Bits<N> and inline if conditions.
Use CSR field access (CSR[name].FIELD) and 1'b1 bit literals.

Shehrozkashif committed:
Merge branch 'main' into H_CSRs

Shehrozkashif committed:
fix(csr): set hgeie/hgeip reset_value to 0 to fix IDL type-checker crash

UNDEFINED_LEGAL caused T.cast failure in Udb::Csr#value when the IDL
type-checker evaluated $bits(CSR[hgeie]) & $bits(CSR[hgeip]) in
hip sw_read. At reset no guest external interrupts are enabled or
pending, so 0 is semantically correct.

Shehrozkashif committed:
Merge branch 'main' into H_CSRs

ThinkOpenly commented on spec/std/isa/csr/H/hgeip.yaml:
```diff
@@ -0,0 +1,33 @@
+    long_name: Guest external interrupt pending bits
+    location_rv32: 31-1
+    location_rv64: 63-1
+    type: RO
```
```suggestion
    type: RO-H
```

ThinkOpenly commented on spec/std/isa/csr/H/hideleg.yaml:
```diff
@@ -0,0 +1,70 @@
+    reset_value: 0
+    long_name: Supervisor Software Interrupt
+    description: Supervisor Software Interrupt
+
```
All of the bits in `hideleg` are gated by their matching bits in `mideleg` and additionally:
> For bits of `mideleg` that are zero, the corresponding bits in `hideleg`, `hip`, and `hie` are read-only zeros.

So, these fields all need `sw_write()` methods that check the respective field in `mideleg` and prevent writing if that field is 0. And, the CSR needs a `sw_read()` field masking by `mideleg`.

ThinkOpenly commented on spec/std/isa/csr/H/hie.yaml:
```diff
@@ -0,0 +1,52 @@
+    name: H
+length: SXLEN
+fields:
+  SGEIE:
```
We'll eventually need to support this:
> For each writable bit in `sie`, the corresponding bit shall be read-only zero in both `hip` and `hie`.

But, I think this is more of a need for parameters which aren't currently defined. So, this comment should be turned into an issue.

ThinkOpenly commented on spec/std/isa/csr/H/vsip.yaml:
```diff
@@ -0,0 +1,77 @@
+      Indicates a pending software interrupt at the supervisor level.
+      Read-only zero when `hideleg[2] == 0`, else aliased to `hip.VSSIP`.
+
+  STIP:
```
Can you add a `sw_write()` method here?

ThinkOpenly commented on spec/std/isa/csr/H/vsip.yaml:
```diff
@@ -0,0 +1,77 @@
+      Indicates a pending timer interrupt at the supervisor level.
+      Read-only zero when `hideleg[6] == 0`, else aliased to `hip.VSTIP`.
+
+  SEIP:
```
Can you add a `sw_write()` method here?

ThinkOpenly commented on spec/std/isa/csr/H/vsscratch.yaml:
```diff
@@ -0,0 +1,29 @@
+fields:
+  SCRATCH:
+    type: RW
+    reset_value: 0
```
I don't see this defined anywhere.
```suggestion
    reset_value: UNDEFINED_LEGAL
```

ThinkOpenly commented on spec/std/isa/csr/stimecmp.yaml:
```diff
@@ -0,0 +1,34 @@
+    location_rv32: 31-0
+    location_rv64: 63-0
+    type: RW
+    reset_value: 0
```
I don't see this defined anywhere. (Seems most logical to me to have it reset to a maximum unsigned value.)
```suggestion
    reset_value: UNDEFINED_LEGAL
```

ThinkOpenly reviewed (changes requested):
Sincere apologies for the LONG delay here. If you can continue here, that's great. If not, just let me know, and we'll get somebody to pick it up from where you left it. We'll keep your commit acknowledgements.
